dynamiqs/_utils.py

```diff
@@ -41,3 +42,27 @@ def bexpect(O: Array, x: Array) -> Array:
     if isket(x):
         return jnp.einsum('ij,bjk,kl->b', dag(x), O, x)  # <x|O|x>
     return jnp.einsum('bij,ji->b', O, x)  # tr(Ox)
+
+
+def compute_batching(f, cartesian_batching, other_in_axes, out_axes, *should_batch):
+    # Build a list of batch mask (ie binary numbers). Each digit is 1 if we
+    # should batch over the dimension and 0 otherwise.
+    # For example if `should_batch` = (True, False, True), the batch_mask will be
+    # [001, 100]
+    batch_mask = [2**k * v for k, v in enumerate(should_batch)]
+
+    # When doing cartesian batching, we need to batch over all dimensions
+    # one by one, but when cartesian batching is off, we want to batch over
+    # every dimension at once, hence the sum.
+    # For example if `should_batch` = (True, False, True), the batch_mask will be
+    # [001, 100] if we want to do cartesian batching and [101] otherwise.
+    if not cartesian_batching:
+        batch_mask = [sum(batch_mask)]
+
+    for mask in reversed(batch_mask):
+        if mask == 0:  # skip the no batching case
+            continue
+        mask = tuple([0 if mask & (1 << i) else None for i in range(len(should_batch))])
+        f = jax.vmap(f, in_axes=mask + other_in_axes, out_axes=out_axes)
+
+    return f
```

Comment:
I like the idea, but the code is very cryptic honestly. What about:
```python
def compute_batching(f, cartesian_batching, is_batched, out_axes):
    if any(is_batched):
        if cartesian_batching:
            idx_batched = np.where(is_batched)[0]
            for i in idx_batched[::-1]:
                in_axes = [None] * len(is_batched)
                in_axes[i] = 0
                f = jax.vmap(f, in_axes=in_axes, out_axes=out_axes)
        else:
            in_axes = list(np.where(is_batched, 0, None))
            f = jax.vmap(f, in_axes=in_axes, out_axes=out_axes)

    return f
```
(with typing + comments) and e.g. for sesolve
```python
    # we vectorize over H and psi0, all other arguments are not vectorized
    is_batched = (H.ndim > 2, psi0.ndim > 2, False, False, False, False, False)
    # the result is vectorized over ysave and Esave
    out_axes = Result(None, None, None, None, 0, 0)
    f = compute_batching(_sesolve, options.cartesian_batching, is_batched, out_axes)
```

Reply:
Or
```
for i, v in enumerate(is_batched[::-1]):
    if v:
        in_axes = [None] * len(is_batched)
        in_axes[i] = 0
        f = jax.vmap(f, in_axes=in_axes[::-1], out_axes=out_axes)
```
for the `cartesian_batching` option.

Reply:
Much better indeed